File: frontend/src/api/recommend.ts
import { api } from "./client";
import type { ApiEnvelope, PaginatedData } from "./client";

export interface RecommendItem {
  type: "fund" | "sector";
  action: "buy" | "add" | "watch" | "stop";
  target_id: string;
  target_name: string;
  target_code: string | null;
  confidence: number;
  reason_summary: string;
  reason_detail: Record<string, unknown> | null;
  risk_warning: string | null;
}

/** 历史推荐记录（带 DB 字段） */
export interface RecommendRecord {
  id: string;
  date: string;
  mode: string;
  type: string;
  action: string;
  target_name: string;
  target_code: string | null;
  confidence: number;
  reason_summary: string;
  reason_detail: Record<string, unknown> | null;
  risk_warning: string | null;
}

/** 子策略中文映射 */
export const MODE_LABELS: Record<string, string> = {
  momentum: "强势延续",
  latent: "低位潜伏",
  rebound: "超跌反弹",
  defensive: "抗跌先锋",
};

/** 子策略描述（tooltip） */
export const MODE_DESCRIPTIONS: Record<string, string> = {
  momentum: "推荐涨幅较大但还有上涨空间的标的",
  latent: "推荐涨幅较小但有潜力的标的",
  rebound: "推荐跌幅较大有望反弹的标的",
  defensive: "推荐跌幅较小但基本面稳健的标的",
};

export async function generateRecommend(params: {
  limit?: number;
  category: string;
  mode: string;
}): Promise<ApiEnvelope<{ items: RecommendItem[]; total: number }>> {
  return api.post("recommend/generate", { json: params }).json();
}

export async function getRecommendHistory(params: {
  page?: number;
  page_size?: number;
  start_date?: string;
  end_date?: string;
  mode?: string;
}): Promise<ApiEnvelope<PaginatedData<RecommendRecord>>> {
  const searchParams = new URLSearchParams();
  if (params.page) searchParams.set("page", String(params.page));
  if (params.page_size) searchParams.set("page_size", String(params.page_size));
  if (params.start_date) searchParams.set("start_date", params.start_date);
  if (params.end_date) searchParams.set("end_date", params.end_date);
  if (params.mode) searchParams.set("mode", params.mode);
  return api.get(`recommend/history?${searchParams}`).json();
}

export async function deleteRecommendHistory(
  ids: string[],
): Promise<ApiEnvelope<{ deleted: number }>> {
  return api.delete(`recommend/history?ids=${ids.join(",")}`).json();
}

File: backend/app/schemas/recommend.py
"""推荐系统 Pydantic 模型."""

from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class RecommendItem(BaseModel):
    """单条推荐结果."""

    type: str = Field(description="推荐类型：fund / sector")
    action: str = Field(
        description="操作建议：buy(推荐)/add(加仓)/watch(观望)/stop(止损)",
    )
    target_id: str = Field(description="目标 ID（基金代码或板块 UUID）")
    target_name: str = Field(description="目标名称")
    target_code: str | None = Field(default=None, description="基金代码或板块代码")
    confidence: int = Field(ge=0, le=100, description="置信度 0-100")
    reason_summary: str = Field(description="推荐理由摘要")
    reason_detail: dict[str, Any] | None = Field(
        default=None, description="详细分析（技术面/基本面/情绪等）"
    )
    risk_warning: str | None = Field(default=None, description="风险提示")
    source_data: dict[str, Any] | None = Field(
        default=None, description="来源数据快照",
    )


class RecommendRequest(BaseModel):
    """推荐请求."""

    limit: int = Field(default=10, ge=1, le=20, description="推荐数量")
    category: str = Field(description="fund / sector")
    mode: str = Field(description="momentum / latent / rebound / defensive")


class DipBuyRequest(BaseModel):
    """加仓推荐请求."""

    limit: int = Field(default=10, ge=1, le=20, description="推荐数量")
    max_drawdown: float = Field(
        default=2.0, ge=0.5, le=50.0, description="阶段跌幅阈值（%），默认 2%",
    )
    min_consecutive_days: int = Field(
        default=2, ge=1, le=20, description="最少连跌天数，默认 2 天",
    )


class RecommendResponse(BaseModel):
    """推荐结果响应."""

    items: list[RecommendItem]
    total: int
